### PR TITLE
fix restoring finetune callbacks after accelerator setup on training resume

### PR DESCRIPTION
**Pull request by eladsegal**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/8502

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
Dear @eladsegal ,

I  have added a test on your branch and some minor fixes. Mind giving it a try to see if it fixes your problem ?

Best,
T.C

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -856,12 +856,13 @@ def _run(self, model: 'pl.LightningModule') -> Optional[Union[_EVALUATE_OUTPUT,
...
         self.checkpoint_connector.restore_model()
-        # restore callback states
-        self.checkpoint_connector.restore_callbacks()
 
         self._call_configure_sharded_model(model)  # allow user to setup in model sharded environment
         self.accelerator.setup(self, model)  # note: this sets up self.lightning_module
 
+        # restore callback states
+        self.checkpoint_connector.restore_callbacks()
+
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:865]: it was changed here: #7652

was the reported issue working fine before this?
ideally, we want to pair `restore_model()` and `restore_callbacks()` together such that the on_load_checkpoint runs right after each other. 

now there are additional hooks running between.
is it not possible to delay optimizer setup in the fine tuning callbacks to `on_train_start` or similar?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:865]: Hey @awaelchli, the test should fail on master due to position of the `setup`. However, if the user was using sharded optimizers, we would need to capture them too.

Comment (eladsegal) [reply pytorch_lightning/trainer/trainer.py:865]: @awaelchli The notebook from the issue https://github.com/PyTorchLightning/pytorch-lightning/issues/8502 ([here](https://colab.research.google.com/drive/1xIE218xOvK9tSoWV_2z820Qog_NqrILs?usp=sharing)) works with version 1.3.8.

**Comment by eladsegal** [comment]
> Dear @eladsegal ,
> 
> I have added a test on your branch and some minor fixes. Mind giving it a try to see if it fixes your problem ?
> 
> Best,
> T.C

Thank you! I was planning on finalizing the pull request, but you were too quick :)
It does solve my problem

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Done !

File: pytorch_lightning/callbacks/finetuning.py [hunk 2]
```diff
@@ -83,6 +83,7 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
 
     def __init__(self):
         self._internal_state: Dict[int, List[Dict[str, Any]]] = {}
+        self.restarting = False
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:86]:
```suggestion
        self._restarting = False
```
rather protected?

File: pytorch_lightning/callbacks/finetuning.py [hunk 3]
```diff
@@ -95,12 +96,17 @@ def on_save_checkpoint(
...
-        named_parameters = dict(pl_module.named_parameters())
-        for opt_idx, optimizer in enumerate(trainer.optimizers):
-            param_groups = self.__apply_mapping_to_param_groups(self._internal_state[opt_idx], named_parameters)
-            optimizer.param_groups = param_groups
+        if self.restarting:
+            named_parameters = dict(pl_module.named_parameters())
+            for opt_idx, optimizer in enumerate(trainer.optimizers):
+                param_groups = self.__apply_mapping_to_param_groups(self._internal_state[opt_idx], named_parameters)
+                optimizer.param_groups = param_groups
+            self.restarting = False
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:109]:
```suggestion
        if not self.restarting:
            return
        named_parameters = dict(pl_module.named_parameters())
        for opt_idx, optimizer in enumerate(trainer.optimizers):
            param_groups = self.__apply_mapping_to_param_groups(self._internal_state[opt_idx], named_parameters)
            optimizer.param_groups = param_groups
        self.restarting = False
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 4]
```diff
@@ -83,6 +83,7 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
 
     def __init__(self):
         self._internal_state: Dict[int, List[Dict[str, Any]]] = {}
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:85]: this is a bit confusing can we rather call it `_internal_optimizer_state` as it it related to optimizer not callback

File: pytorch_lightning/callbacks/finetuning.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:86]: Done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py]: Done !

**Comment by awaelchli** [comment]
I'm changing the title of this PR to reflect that this is about the finetuning callback implementation :) 😇

File: pytorch_lightning/callbacks/finetuning.py [hunk 5]
```diff
@@ -82,25 +82,33 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
...
+        return self._internal_optimizer_metadata
 
     def on_load_checkpoint(
         self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule', callback_state: Dict[int, List[Dict[str, Any]]]
     ) -> None:
-        self._internal_state = callback_state
+        self.restarting = True
+        self._internal_optimizer_metadata = callback_state
+
+    def on_fit_start(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule') -> None:
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/finetuning.py:102]: great! glad this worked for you :) 
<3

File: pytorch_lightning/callbacks/finetuning.py [hunk 6]
```diff
@@ -362,15 +372,33 @@ def __init__(
...
+        self.unfreeze_backbone_at_epoch: int = unfreeze_backbone_at_epoch
+        self.backbone_initial_lr: Optional[float] = backbone_initial_lr
+        self.lambda_func: Callable = lambda_func
+        self.backbone_initial_ratio_lr: float = backbone_initial_ratio_lr
+        self.should_align: bool = should_align
+        self.initial_denom_lr: float = initial_denom_lr
+        self.previous_backbone_lr: Optional[float] = None
+        self.train_bn: bool = train_bn
+        self.round: int = round
+        self.verbose: bool = verbose
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/finetuning.py:384]: nit: order of these is not the same as in the init signature :)

File: tests/callbacks/test_finetuning_callback.py [hunk 7]
```diff
@@ -365,3 +365,118 @@ def forward(self, x):
...
+    # Initialize a trainer
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        max_epochs=3,
+        progress_bar_refresh_rate=0,
+        callbacks=[cb, BackboneFinetuning(unfreeze_backbone_at_epoch=2)]
+    )
+
+    # Train the model ⚡
```
Comment (awaelchli) [inline tests/callbacks/test_finetuning_callback.py:470]: hehe, taken straight from the examples?
we should consider cleaning these comments out, as the code is pretty verbose in what it is doing, so these comments are not necessary :)

**Review by awaelchli — APPROVED** [review]
LGTM overall 
thanks @eladsegal and @tchaton for taking care of my negligence in #7652, I have not paid enough attention and should have seen that it breaks

File: tests/callbacks/test_finetuning_callback.py [hunk 8]
```diff
@@ -365,3 +365,118 @@ def forward(self, x):
...
+        def forward(self, x):
+            return self.layer(self.backbone(x))
+
+    cb = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+
+    # Initialize a trainer
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        max_epochs=3,
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:465]: Does the test need 3 epochs?

File: tests/callbacks/test_finetuning_callback.py [hunk 9]
```diff
@@ -365,3 +365,118 @@ def forward(self, x):
...
+
+        def forward(self, x):
+            return self.layer(self.backbone(x))
+
+    cb = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+
+    # Initialize a trainer
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:464]: Also limit the number of val batches?

File: tests/callbacks/test_finetuning_callback.py [hunk 10]
```diff
@@ -365,3 +365,118 @@ def forward(self, x):
...
+    )
+
+    # Train the model ⚡
+    trainer.fit(BackboneBoringModel())
+
+    # Initialize a trainer that continues the previous training
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        max_epochs=4,
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:477]: Same coment

File: tests/callbacks/test_finetuning_callback.py [hunk 11]
```diff
@@ -365,3 +365,118 @@ def forward(self, x):
...
+    """
+
+    seed_everything(42)
+
+    chk = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+
+    model = FinetuningBoringModel()
+    callback = TestCallbacksRestoreCallback()
+
+    trainer_kwargs = dict(limit_train_batches=1, default_root_dir=tmpdir, callbacks=[callback, chk], max_epochs=2)
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:405]: same comments

File: tests/callbacks/test_finetuning_callback.py [hunk 12]
```diff
@@ -365,3 +365,118 @@ def forward(self, x):
...
+        return optimizer
+
+
+def test_callbacks_restore(tmpdir):
+    """
+    Test callbacks restore is called after optimizers have been re-created
+    but before optimizer states reload
+    """
+
+    seed_everything(42)
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:398]: Is the fixed seed strictly necessary for the test?

File: tests/callbacks/test_finetuning_callback.py (hunk 8 above)
Comment (awaelchli) [reply tests/callbacks/test_finetuning_callback.py]: nope, only requirement is that the second trainer has one epoch more .)

File: tests/callbacks/test_finetuning_callback.py (hunk 12 above)
Comment (awaelchli) [reply tests/callbacks/test_finetuning_callback.py]: addressed all your comments @carmocca

File: tests/callbacks/test_finetuning_callback.py [hunk 13]
```diff
@@ -365,3 +365,115 @@ def forward(self, x):
...
+            return self.layer(self.backbone(x))
+
+    cb = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        limit_val_batches=1,
+        max_epochs=1,
+        progress_bar_refresh_rate=0,
+        callbacks=[cb, BackboneFinetuning(unfreeze_backbone_at_epoch=2)]
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:465]: @awaelchli I think you forgot to update these. This should make the test fail

Comment (awaelchli) [reply tests/callbacks/test_finetuning_callback.py:465]: hmm, this test doesn't have any assertions. I don't know what is supposed to fail but I guess the finetuning callback doesn't care if we never reach the epoch at which we unfreeze. 

Let's revert back to 2 and 3 epochs respectively?

File: tests/callbacks/test_finetuning_callback.py [hunk 14]
```diff
@@ -365,3 +365,115 @@ def forward(self, x):
...
+    trainer_kwargs["resume_from_checkpoint"] = chk.last_model_path
+
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model)
+
+
+def test_callbacks_restore_backbone(tmpdir):
+    """
+    Test callbacks restore is called after optimizers have been re-created
+    but before optimizer states reload
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:445]: This test doesn't fail in master, so it does not seem to be testing anything

cc @tchaton who added it

**Review by carmocca — CHANGES_REQUESTED** [review]

File: tests/callbacks/test_finetuning_callback.py (hunk 14 above)
Comment (awaelchli) [reply tests/callbacks/test_finetuning_callback.py:445]: I will revisit the reported failure case

File: tests/callbacks/test_finetuning_callback.py [hunk 15]
```diff
@@ -365,3 +365,115 @@ def forward(self, x):
...
+            return self.layer(self.backbone(x))
+
+    cb = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        limit_val_batches=1,
+        max_epochs=2,
+        progress_bar_refresh_rate=0,
+        callbacks=[cb, BackboneFinetuning(unfreeze_backbone_at_epoch=1)]
```
Comment (awaelchli) [inline tests/callbacks/test_finetuning_callback.py:465]: @carmocca in order to provoke this test to fail on master, the max_epochs here needs to be a larger number than unfreeze_backbone_at_epoch. So I adjusted it.

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_finetuning_callback.py [hunk 16]
```diff
@@ -365,3 +365,115 @@ def forward(self, x):
...
+    trainer.fit(BackboneBoringModel())
+
+    # initialize a trainer that continues the previous training
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        limit_val_batches=1,
+        max_epochs=3,
+        progress_bar_refresh_rate=0,
+        callbacks=BackboneFinetuning(unfreeze_backbone_at_epoch=1),
```
Comment (Borda) [inline tests/callbacks/test_finetuning_callback.py:476]:
```suggestion
        max_epochs=trainer.current_epoch + 1,
        callbacks=BackboneFinetuning(),
```
all this shall be restored

Comment (awaelchli) [reply tests/callbacks/test_finetuning_callback.py:476]: the `unfreeze_backbone_at_epoch` attribute does not get restored. I guess this is done so that we can choose a different value for another fine tuning. 

and all the other trainer properties don't get restored.

**Review by Borda — APPROVED** [review]

File: tests/callbacks/test_finetuning_callback.py (hunk 16 above)
Comment (tchaton) [reply tests/callbacks/test_finetuning_callback.py:476]: Hey @awaelchli, I believe we didn't dive too deep into restoring Fine-tuning Callback yet. 
This can be seen as a bug or feature to enable users to provide different arguments.
Curious to hear your thoughts.
